docs/data.js

```diff
@@ -0,0 +1,93 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Adds an array of strings to a set individually.
+ *
+ * @param {Set<string>} set
+ * @param {string[]} add
+ */
+const addToSet = (set, add) => {
+    for (item of add) {
+        set.add(item);
+    }
+}
+
+/**
+ * Calculate the difference of two sets. Implementation is from
+ * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set#implementing_basic_set_operations
+ * @param {Set} setA
+ * @param {Set} setB
+ */
+function difference(setA, setB) {
+    const _difference = new Set(setA);
+    for (const elem of setB) {
+        _difference.delete(elem);
+    }
+    return _difference;
+}
+
+// exports.addToSet = addToSet;
+const memberCombineInstructions = [
+    {
+        package: "@fluidframework/azure-service-utils",
+        sourceImports: new Map([
+            ["@fluidframework/server-services-client", ["GenerateToken"]],
+            ["@fluidframework/protocol-definitions", ["ScopeType"]],
+        ])
+    },
+    {
+        package: "@fluidframework/fluid-static",
+        sourceImports: new Map([
+            ["@fluidframework/container-definitions", ["AttachState", "IAudience"]]
+        ])
+    },
+];
+exports.memberCombineInstructions = memberCombineInstructions;
+
+const packageRollupMap = new Map([
+    // fluid-framework re-exports all of fluid-static
+    ["fluid-framework", [
```

Comment:
what's the intent around the `container-definitions` re-exports?  if i'm understanding correctly, the `memberCombineInstructions` above is attributing them to `fluid-static`, but `fluid-framework` is the one that's exporting them.

Reply:
The short answer is that the input data is structured this way because that's how we actually re-export things in the packages: `AttachState` from `container-definitions` is imported by `fluid-static`, then re-exported, and then re-exported _again_ by `fluid-framework`. My goal was to structure the input data as closely as possible to the way the re-exported APIs flow through the packages. My reasoning was twofold:

1. It's easier to understand what the script is doing if it processes things in the same way the code does with its import/exports.
2. We could in the future source the input data from the ts files themselves by parsing them into an AST. This approach makes it easy to swap out the input data source without also doing some pre-optimization on it.

That said, your comment made me think about another potential optimization: if I change the `memberCombineInstructions` handling to understand importing all of a package, I could probably get rid of the rollup map altogether. E.g. something like this:

```ts
const memberCombineInstructions = [
    {
        package: "@fluidframework/azure-service-utils",
        sourceImports: new Map([
            ["@fluidframework/server-services-client", ["GenerateToken"]],
            ["@fluidframework/protocol-definitions", ["ScopeType"]],
        ])
    },
    {
        package: "@fluidframework/fluid-static",
        sourceImports: new Map([
            ["@fluidframework/container-definitions", ["AttachState", "IAudience"]]
        ])
    },
    {
        package: "fluid-framework",
        sourceImports: new Map([
            ["@fluidframework/fluid-static", ["*"]],
            ["@fluidframework/map", ["*"]],
            ["@fluidframework/sequence", ["*"]],
        ])
    },
];
```

@heliocliu Do you think that would make the code more or less clear?

Reply:
I think getting rid of the rollup map would be a lot clearer, since having both means you need to understand how they're different (when they seem very similar).

As for `AttachState`, it's kind of a weird situation because `fluid-static` doesn't explicitly export it, but rather as part of another type.  So someone directly consuming `fluid-static` wouldn't be able to tell its definition without also importing `container-definitions`.  Applying the API flow logic to the rest of the runtime packages, I'd guess you'd see a lot of unexpected attributions (or maybe they're expected and just more difficult to reason about because they become very distributed).  Then in just the scope of the 3P libraries, if a parallel library to `fluid-static` were added and also exported the same variable in the same way, would that variable be attributed to both these libraries?

Ultimately I don't think it's a big deal either way as long as this is all scoped to just the 3p packages, so if one makes future work easier then that seems fine to me.

Reply:
> I think getting rid of the rollup map would be a lot clearer, since having both means you need to understand how they're different (when they seem very similar).

The latest change removes the package map and replaces all its logic with some special handling for `"*"` in the combineMemberInstructions.

> As for AttachState, it's kind of a weird situation because fluid-static doesn't explicitly export it, but rather as part of another type.

Good point on `AttachState`! It was a bug that I thought AttachState was also exported from fluid-static. The latest change splits them:

```ts
const memberCombineInstructions = [
    {
        package: "@fluidframework/azure-service-utils",
        sourceImports: new Map([
            // TODO: #7530
            // ["@fluidframework/server-services-client", ["GenerateToken"]],
            ["@fluidframework/protocol-definitions", ["ScopeType"]],
        ])
    },
    {
        package: "@fluidframework/fluid-static",
        sourceImports: new Map([
            ["@fluidframework/container-definitions", ["IAudience"]],
        ])
    },
    {
        package: "fluid-framework",
        sourceImports: new Map([
            ["@fluidframework/container-definitions", ["AttachState"]],
            ["@fluidframework/fluid-static", ["*"]],
            ["@fluidframework/map", ["*"]],
            ["@fluidframework/sequence", ["*"]],
        ])
    },
];
```

Reply:
Well in that case `IAudience` is in the same situation of not being explicitly exported 🤔